File: app/src/main/java/co/appdev/boilerplate/injection/component/ApplicationComponent.java
package co.appdev.boilerplate.injection.component;

import javax.inject.Singleton;

import co.appdev.boilerplate.data.DataManager;
import co.appdev.boilerplate.data.local.DatabaseRealm;
import co.appdev.boilerplate.data.local.SharedPrefHelper;
import co.appdev.boilerplate.injection.ApplicationContext;
import co.appdev.boilerplate.injection.module.ApplicationModule;
import co.appdev.boilerplate.injection.module.NetworkModule;
import co.appdev.boilerplate.injection.module.SharedPreferencesModule;
import co.appdev.boilerplate.services.SyncService;
import co.appdev.boilerplate.ui.base.BaseActivity;
import co.appdev.boilerplate.ui.base.BaseFragment;
import co.appdev.boilerplate.ui.splash.SplashActivity;
import dagger.Component;

@Singleton
@Component(modules = {ApplicationModule.class,NetworkModule.class, SharedPreferencesModule.class})
public interface ApplicationComponent {

    @ApplicationContext
    DataManager dataManager();
    SharedPrefHelper sharedPrefHelper();
    DatabaseRealm databaseRealm();

    void inject(SyncService syncService);
    void inject(BaseFragment baseFragment);
    void inject(BaseActivity baseActivity);
    void inject(SplashActivity splashActivity);
}

File: app/src/main/java/co/appdev/boilerplate/di/component/ApplicationComponent.kt
package co.appdev.boilerplate.di.component

import co.appdev.boilerplate.data.DataManager
import co.appdev.boilerplate.data.local.SharedPrefHelper
import co.appdev.boilerplate.di.module.ApplicationModule
import co.appdev.boilerplate.di.module.NetworkModule
import co.appdev.boilerplate.di.module.RoomModule
import co.appdev.boilerplate.di.module.SharedPreferencesModule
import co.appdev.boilerplate.services.SyncService
import co.appdev.boilerplate.ui.list.UserPostListViewModel
import co.appdev.boilerplate.ui.list.UserPostViewModel
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [NetworkModule::class, ApplicationModule::class, RoomModule::class, SharedPreferencesModule::class])
interface ApplicationComponent {

    fun dataManager(): DataManager

    /**
     * Injects required dependencies into the specified PostListViewModel.
     * @param postListViewModel PostListViewModel in which to inject the dependencies
     */
    fun inject(postListViewModel: UserPostListViewModel)

    /**
     * Injects required dependencies into the specified PostViewModel.
     * @param postViewModel PostViewModel in which to inject the dependencies
     */
    fun inject(postViewModel: UserPostViewModel)

    fun inject(syncService: SyncService)


}

